### PR TITLE
Messaging

### PR DESCRIPTION
**Pull request by KrauseFx**
Writing a new abstraction for printing out messages

This will allow us to have different implementations in the future, and not only the terminal.

### CODE REVIEW

File: lib/fastlane_core/command_executor.rb [hunk 1]
```diff
@@ -54,7 +52,7 @@ def execute(command: nil, print_all: false, print_command: true, error: nil, pre
         if status != 0
           o = output.join("\n")
           puts o # the user has the right to see the raw output
-          Helper.log.info "Exit status: #{status}"
+          UI.error "Exit status: #{status}"
```
Comment (jcampbell05) [inline lib/fastlane_core/command_executor.rb]: Nit picky but how come in this use case it isn't surrounded with parenthesis

File: lib/fastlane_core/ui/implementations/shell.rb [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+      log.info(message)
+    end
+
+    def command(message)
+      log.info("$ #{message}".cyan.underline)
+    end
+
+    def command_output(message)
+      actual = (message.split("\r").last || "") # as clearing the line will remove the `>` and the time stamp
+      actual.split("\n").each do |msg|
```
Comment (mfurtak) [inline lib/fastlane_core/ui/implementations/shell.rb]: A future concern perhaps, but will these `split`s be correct for line endings on other OSes?

Comment (KrauseFx) [inline lib/fastlane_core/ui/implementations/shell.rb]: @mfurtak Do you mean e.g. Windows which doesn't use `\n`? Do you know if there is a general way to split per line to work on all OSes?

Comment (mfurtak) [inline lib/fastlane_core/ui/implementations/shell.rb]: Right. That's the sort of thing I am thinking about, though I admit that it is probably premature to worry about in this PR.

In very quick research `s.encode(encoding, universal_newline: true)` might be the way to normalize, but I'm sure others with more Ruby experience than me can say better!

**Comment by mfurtak** [comment]
Aside from noted future concerns, I think this is looking great. Squash and :ship: it!

**Comment by KrauseFx** [comment]
Great, thank you for @mfurtak for the fast review :rocket:

**Pull request merged**